File: src/application/catalog.ts
import fs from "node:fs";
import path from "node:path";
import {
  asBoolean01,
  asString,
  asStringArray,
  type FrontmatterData,
} from "../domain/frontmatter.js";
import {
  ENTITY_TYPES,
  type EntityType,
} from "../domain/entities.js";
import {
  listEntityFiles,
  type EntityFile,
} from "../infrastructure/entities.js";

export type CatalogEntry = {
  id: string;
  type: EntityType;
  path: string;
  title: string;
  status: string;
  mtimeMs: number;
  data: FrontmatterData;
  /** Body not loaded by default (frontmatter-only scan). */
  bodyPreview?: string;
};

export type ProjectCatalog = {
  projectRoot: string;
  entries: CatalogEntry[];
  byId: Map<string, CatalogEntry[]>;
  byType: Record<EntityType, CatalogEntry[]>;
};

function entryTitle(type: EntityType, data: FrontmatterData, id: string): string {
  if (type === "intake" || type === "report") {
    return asString(data, "summary") ?? id;
  }
  return asString(data, "title") ?? id;
}

function entryStatus(type: EntityType, data: FrontmatterData): string {
  return asString(data, "status") ?? "";
}

export function fileToCatalogEntry(file: EntityFile): CatalogEntry {
  const id = asString(file.data, "id") ?? path.basename(file.relativePath, ".md");
  const type = (asString(file.data, "type") as EntityType) ?? "story";
  let mtimeMs = 0;
  try {
    mtimeMs = fs.statSync(file.absolutePath).mtimeMs;
  } catch {
    mtimeMs = 0;
  }
  return {
    id,
    type,
    path: file.relativePath.replace(/\\/g, "/"),
    title: entryTitle(type, file.data, id),
    status: entryStatus(type, file.data),
    mtimeMs,
    data: file.data,
  };
}

/**
 * Scan entity markdown directories (frontmatter only via listEntityFiles).
 * Shared by query (US-008) and index (US-009).
 */
export function buildCatalog(projectRoot: string): ProjectCatalog {
  const entries: CatalogEntry[] = [];
  const byType = {
    story: [] as CatalogEntry[],
    decision: [] as CatalogEntry[],
    intake: [] as CatalogEntry[],
    backlog: [] as CatalogEntry[],
    report: [] as CatalogEntry[],
  };

  for (const type of ENTITY_TYPES) {
    const files = listEntityFiles(projectRoot, type);
    for (const file of files) {
      const entry = fileToCatalogEntry(file);
      // normalize type from path when frontmatter omitted type but was filtered
      if (!ENTITY_TYPES.includes(entry.type)) {
        entry.type = type;
      }
      entries.push(entry);
      byType[type].push(entry);
    }
  }

  entries.sort((a, b) => a.id.localeCompare(b.id));
  for (const t of ENTITY_TYPES) {
    byType[t].sort((a, b) => a.id.localeCompare(b.id));
  }

  const byId = new Map<string, CatalogEntry[]>();
  for (const e of entries) {
    const list = byId.get(e.id) ?? [];
    list.push(e);
    byId.set(e.id, list);
  }

  return { projectRoot, entries, byId, byType };
}

export function resolveCatalogEntry(
  catalog: ProjectCatalog,
  idOrPath: string,
): CatalogEntry | null {
  const raw = idOrPath.trim().replace(/\\/g, "/");
  if (!raw) return null;

  // exact path match
  const byPath = catalog.entries.find(
    (e) => e.path === raw || e.path === raw.replace(/^\.\//, ""),
  );
  if (byPath) return byPath;

  // type/id path fragment
  const slash = raw.includes("/") ? raw : null;
  if (slash) {
    const base = path.posix.basename(raw, ".md");
    const hits = catalog.byId.get(base) ?? [];
    const typed = hits.find(
      (h) =>
        h.path.endsWith(`/${base}.md`) &&
        (raw.includes(h.type) || raw.includes(path.posix.dirname(h.path))),
    );
    if (typed) return typed;
    if (hits.length === 1) return hits[0]!;
  }

  const hits = catalog.byId.get(raw) ?? [];
  if (hits.length === 1) return hits[0]!;
  if (hits.length > 1) {
    // prefer story for bare US-* ids
    const story = hits.find((h) => h.type === "story");
    if (story) return story;
    return hits[0]!;
  }
  return null;
}

export function proof01(data: FrontmatterData, key: string): 0 | 1 {
  return asBoolean01(data, key) ?? 0;
}

export function linksOf(data: FrontmatterData): string[] {
  return [
    ...(asStringArray(data, "links") ?? []),
    ...(asStringArray(data, "related") ?? []),
  ].filter((value, index, all) => all.indexOf(value) === index);
}

File: src/app/catalog.rs
use std::path::Path;

use crate::domain::entities::ENTITY_TYPES;
use crate::domain::frontmatter::{as_bool01, as_string, as_string_array, Frontmatter};
use crate::error::Result;
use crate::infra::entities::{list_entity_files, EntityFile};

#[derive(Clone, Debug)]
pub struct CatalogEntry {
    pub id: String,
    pub ty: String,
    pub path: String,
    pub title: String,
    pub status: String,
    pub mtime_ms: u128,
    pub data: Frontmatter,
}

pub struct ProjectCatalog {
    pub entries: Vec<CatalogEntry>,
}

pub fn file_to_catalog_entry(file: &EntityFile) -> CatalogEntry {
    let id = as_string(&file.data, "id").unwrap_or_else(|| {
        Path::new(&file.relative_path)
            .file_stem()
            .map(|s| s.to_string_lossy().into_owned())
            .unwrap_or_default()
    });
    let ty = as_string(&file.data, "type").unwrap_or_else(|| "story".into());
    let title = if ty == "intake" || ty == "report" {
        as_string(&file.data, "summary").unwrap_or_else(|| id.clone())
    } else {
        as_string(&file.data, "title").unwrap_or_else(|| id.clone())
    };
    let mtime_ms = file
        .absolute_path
        .metadata()
        .and_then(|m| m.modified())
        .ok()
        .and_then(|t| t.duration_since(std::time::UNIX_EPOCH).ok())
        .map(|d| d.as_millis())
        .unwrap_or(0);
    CatalogEntry {
        id,
        ty: ty.clone(),
        path: file.relative_path.replace('\\', "/"),
        title,
        status: as_string(&file.data, "status").unwrap_or_default(),
        mtime_ms,
        data: file.data.clone(),
    }
}

pub fn build_catalog(project_root: &Path) -> Result<ProjectCatalog> {
    let mut entries = Vec::new();
    for ty in ENTITY_TYPES {
        for file in list_entity_files(project_root, ty)? {
            let mut entry = file_to_catalog_entry(&file);
            if !ENTITY_TYPES.contains(&entry.ty.as_str()) {
                entry.ty = (*ty).to_string();
            }
            entries.push(entry);
        }
    }
    entries.sort_by(|a, b| a.id.cmp(&b.id));
    Ok(ProjectCatalog { entries })
}

pub fn by_type<'a>(catalog: &'a ProjectCatalog, ty: &str) -> Vec<&'a CatalogEntry> {
    catalog.entries.iter().filter(|e| e.ty == ty).collect()
}

pub fn proof01(data: &Frontmatter, key: &str) -> i64 {
    as_bool01(data, key).unwrap_or(0)
}

pub fn links_of(data: &Frontmatter) -> Vec<String> {
    as_string_array(data, "links").unwrap_or_default()
}
